src/GafferScene/ResamplePrimitiveVariables.cpp

```diff
@@ -0,0 +1,119 @@
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "boost/format.hpp"
+#include "boost/array.hpp"
+
+#include "GafferScene/ResamplePrimitiveVariables.h"
+
+#include "IECore/MeshPrimitive.h"
+#include "IECore/CurvesPrimitive.h"
+#include "IECore/PointsPrimitive.h"
+#include "IECore/MeshAlgo.h"
+#include "IECore/CurvesAlgo.h"
+#include "IECore/PointsAlgo.h"
+
+using namespace IECore;
+using namespace Gaffer;
+using namespace GafferScene;
+
+IE_CORE_DEFINERUNTIMETYPED( ResamplePrimitiveVariables );
+
+size_t ResamplePrimitiveVariables::g_firstPlugIndex = 0;
+
+ResamplePrimitiveVariables::ResamplePrimitiveVariables( const std::string &name ) : PrimitiveVariableProcessor( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+
+	addChild( new IntPlug( "interpolation" ) );
+}
+
+ResamplePrimitiveVariables::~ResamplePrimitiveVariables()
+{
+}
+
+Gaffer::IntPlug *ResamplePrimitiveVariables::interpolationPlug()
+{
+	return getChild<IntPlug>( g_firstPlugIndex );
+}
+
+const Gaffer::IntPlug *ResamplePrimitiveVariables::interpolationPlug() const
+{
+	return getChild<IntPlug>( g_firstPlugIndex );
+}
+
+void ResamplePrimitiveVariables::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const
+{
+	PrimitiveVariableProcessor::affects( input, outputs );
+
+	if( input == interpolationPlug() )
+	{
+		outputs.push_back( outPlug()->objectPlug() );
+	}
+}
+
+void ResamplePrimitiveVariables::hashProcessedObject( const ScenePath &path, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	PrimitiveVariableProcessor::hashProcessedObject( path, context, h );
+
+	interpolationPlug()->hash( h );
+}
+
+void ResamplePrimitiveVariables::processPrimitiveVariable( const ScenePath &path, const Gaffer::Context *context, IECore::ConstPrimitivePtr inputGeometry, IECore::PrimitiveVariable &variable ) const
+{
+	static boost::array<PrimitiveVariable::Interpolation, 4> mapping = {
+		PrimitiveVariable::Constant, PrimitiveVariable::Uniform, PrimitiveVariable::Vertex, PrimitiveVariable::FaceVarying
+	};
+
+	PrimitiveVariable::Interpolation interpolation = mapping.at( interpolationPlug()->getValue() ); // throws if our interpolation plug value isn't a valid interpolation enum.
+
+	if( ConstMeshPrimitivePtr meshPrimitive = IECore::runTimeCast<const MeshPrimitive>( inputGeometry ) )
```

Comment:
This is probably a minor thing in this context, but since ownership is maintained by our `IECore::ConstPrimitivePtr inputGeometry`, we can use raw pointers here and avoid some unnecessary increments and decrements of the reference count (which because it is atomic, does incur some synchronisation cost) :

`const MeshPrimitive *meshPrimitive = IECore::runTimeCast<const MeshPrimitive>( inputGeometry.get() )`